extern/sector-storage/sched.go

```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
-	var ready sync.WaitGroup
-	ready.Add(1)
-	defer ready.Wait()
+// context only used for startup
+func (sh *scheduler) runWorker(ctx context.Context, w Worker) error {
+	info, err := w.Info(ctx)
+	if err != nil {
+		return xerrors.Errorf("getting worker info: %w", err)
+	}
 
-	go func() {
-		sh.workersLk.RLock()
-		worker, found := sh.workers[wid]
-		sh.workersLk.RUnlock()
+	sessID, err := w.Session(ctx)
+	if err != nil {
+		return xerrors.Errorf("getting worker session: %w", err)
+	}
+	if sessID == ClosedWorkerID {
+		return xerrors.Errorf("worker already closed")
+	}
 
-		ready.Done()
+	worker := &workerHandle{
+		w:    w,
+		info: info,
 
-		if !found {
-			panic(fmt.Sprintf("worker %d not found", wid))
-		}
+		preparing: &activeResources{},
+		active:    &activeResources{},
+		enabled:   true,
+
+		closingMgr: make(chan struct{}),
+		closedMgr:  make(chan struct{}),
+	}
+
+	wid := WorkerID(sessID)
+
+	sh.workersLk.Lock()
+	_, exist := sh.workers[wid]
+	if exist {
+		// this is ok, we're already handling this worker in a different goroutine
+		return nil
+	}
+
+	sh.workers[wid] = worker
+	sh.workersLk.Unlock()
+
+	go func() {
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
 
 		defer close(worker.closedMgr)
 
 		scheduledWindows := make(chan *schedWindow, SchedWindows)
 		taskDone := make(chan struct{}, 1)
 		windowsRequested := 0
 
-		ctx, cancel := context.WithCancel(context.TODO())
-		defer cancel()
+		disable := func(ctx context.Context) error {
+			done := make(chan struct{})
 
-		workerClosing, err := worker.w.Closing(ctx)
-		if err != nil {
-			return
+			// request cleanup in the main scheduler goroutine
+			select {
+			case sh.workerDisable <- workerDisableReq{
+				activeWindows: worker.activeWindows,
+				wid:           wid,
+				done: func() {
+					close(done)
+				},
+			}:
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-sh.closing:
+				return nil
+			}
+
+			// wait for cleanup to complete
+			select {
+			case <-done:
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-sh.closing:
+				return nil
+			}
+
+			worker.activeWindows = worker.activeWindows[:0]
+			windowsRequested = 0
+			return nil
 		}
 
 		defer func() {
-			log.Warnw("Worker closing", "workerid", wid)
+			log.Warnw("Worker closing", "workerid", sessID)
 
-			// TODO: close / return all queued tasks
+			if err := disable(ctx); err != nil {
+				log.Warnw("failed to disable worker", "worker", wid, "error", err)
+			}
+
+			sh.workersLk.Lock()
+			delete(sh.workers, wid)
+			sh.workersLk.Unlock()
 		}()
 
+		heartbeatTimer := time.NewTicker(stores.HeartbeatInterval)
+		defer heartbeatTimer.Stop()
+
 		for {
-			// ask for more windows if we need them
-			for ; windowsRequested < SchedWindows; windowsRequested++ {
+			sh.workersLk.Lock()
+			enabled := worker.enabled
+			sh.workersLk.Unlock()
+
+			// ask for more windows if we need them (non-blocking)
+			for ; enabled && windowsRequested < SchedWindows; windowsRequested++ {
```

Comment:
the nesting here is really painful. Can we maybe break this up? (can be in a followup PR, but should happen soon to make this logic more readable IMO)